File: Data Structures & Algorithms/three-integer-sum/submission-4.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    threeSum(nums) {
        // sort nums
        nums.sort((a,b)=>a-b);

        // initialize result array
        let result = [];

        let n = nums.length;
        
        // loop over nums and for each element,
        for(let i=0;i<n;i++){
            if(i > 0 && nums[i] == nums[i-1]) continue;
            // set target = 0 - nums[i]
            let target = 0 - nums[i];
            // initialize pointers start = i+1 and end = n-1
            let start = i+1, end = n-1;
            // run loop while start<end
            while(start < end){
                let sum = nums[start]+nums[end]
                // if sum(start+end) value > target, end--
                if(sum > target) end--
                // else if sum < target, start++
                else if(sum < target) start++
                // else (sum = target), add to result
                else {
                    result.push([nums[i], nums[start],nums[end]])
                    start++; 
                    end--;
                    while(start < end && nums[start] == nums[start-1]) start++;
                }
            }
        }
        // return result array
        return result;
    }
}

File: Data Structures & Algorithms/longest-consecutive-sequence/submission-4.cpp
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        // sort input array
        // 2,3,4,4,5,10,20
        if(nums.size()==0) return 0;

        int result = 0;
        sort(nums.begin(),nums.end());

        int tempCount = 1;
        // loop over input array
        for(int i=1;i<nums.size();i++){
            // nums[i] = 1 + nums[i-1] -> in sequence, result+1
            if(nums[i] == nums[i-1]+1){
                tempCount++;
            }
            // nums[i] = nums[i-1] -> skip
            else if(nums[i] == nums[i-1]){
                continue;
            }
            // nums[i] > 1+nums[i-1] -> not in sequence, reset result
            else if(nums[i] > 1+nums[i-1]){
                result = max(result,tempCount);
                tempCount = 1;
            }
        }
        result = max(result,tempCount);

        // return result
        return result;
    }
};
